media-proxy/include/mesh/st2110rx.h

```diff
@@ -0,0 +1,138 @@
+#ifndef ST2110RX_H
+#define ST2110RX_H
+
+#include "st2110.h"
+
+namespace mesh::connection {
+
+/**
+ * ST2110Rx
+ *
+ * Base abstract class of ST2110Rx. ST2110_20Rx/ST2110_22Rx/ST2110_30Rx
+ * inherit this class.
+ */
+template <typename FRAME, typename HANDLE, typename OPS> class ST2110Rx : public ST2110 {
+  public:
+    ST2110Rx()
+    {
+        _kind = Kind::receiver;
+
+        _handle = nullptr;
+        _ops = {0};
+        _transfer_size = 0;
+    }
+    ~ST2110Rx() { shutdown(_ctx); }
+
+  protected:
+    HANDLE _handle;
+    OPS _ops;
+    size_t _transfer_size;
+    std::jthread _frame_thread_handle;
+    context::Context _ctx;
+
+    std::function<FRAME *(HANDLE)> _get_frame_fn;
```

Comment:
Why this isn't just a virtual function which is overwritten by concrete function in ST2110_**Rx?

Reply:
I think that dynamic dispatch will do its magic and correct function will be called in e.g.  on_establish()

Reply:
Is it possible to overwrite virtual function with library API call ?
_get_frame_fn points to MTL API call st20p_rx_get_frame(), st22p_rx_get_frame(), st30p_rx_get_frame()

Reply:
I think it isn't possible to do it in runtime, but do you need that? I see that you assign functions to those function pointers in constructors, so it looks like you are doing the job of c++ polymorphism by hand.

If I understand the code correctly you anyway will create objects of derived class, like ST2110_20Rx, so _get_frame_fn (the virtual method in this case) will point to the correct concrete method.